Strangely, on my machine, `test model --max-trace <num>` non-deterministically fails on the _second generated test_(in the loop) with the following error,
```
ImportError while importing test module '/home/ranadeep/informal/atomkraft/examples/cosmos-sdk/transfer/transfer/tests/test_transfer_2022_09_05T17_17_33_148/test_Ex_violation2.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
/usr/lib/python3.10/importlib/__init__.py:126: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
E   ModuleNotFoundError: No module named 'test_Ex_violation2'
```
Not sure what's causing it.

--- a/atomkraft/model/traces.py
+++ b/atomkraft/model/traces.py
@@ -1,5 +1,5 @@
 from pathlib import Path
-from typing import List, Optional
+from typing import Dict, List, Optional
 
 from atomkraft.config.atomkraft_config import AtomkraftConfig
 from atomkraft.config.model_config import ModelConfig
@@ -22,7 +22,7 @@ def generate_traces(
     model_config_path: Optional[Path],
     model_path: Optional[Path] = None,
     sample_operators=[],
-    max_trace: Optional[int] = None,
+    checker_params: Optional[Dict[str, str]] = None,
 ) -> ModelResult:
     """
     Call Modelator to get samples of the given model in `model_path`. Return the
@@ -55,9 +55,8 @@ def generate_traces(
     if not model_path.is_file():
         raise FileNotFoundError(f"File with model not found: {model_path}")
 
-    checker_params = dict()
-    if max_trace:
-        checker_params["max_error"] = str(max_trace)
+    if checker_params is None:
+        checker_params = dict()
 
     model = Model.parse_file(str(model_path), init, next)
     return model.sample(

--- a/atomkraft/model/traces.py
+++ b/atomkraft/model/traces.py
@@ -22,7 +22,7 @@ def generate_traces(
     model_config_path: Optional[Path],
     model_path: Optional[Path] = None,
     sample_operators=[],
-    max_traces: Optional[int] = None,
+    max_trace: Optional[int] = None,
 ) -> ModelResult:
     """
     Call Modelator to get samples of the given model in `model_path`. Return the
@@ -56,8 +56,8 @@ def generate_traces(
         raise FileNotFoundError(f"File with model not found: {model_path}")
 
     checker_params = dict()
-    if max_traces:
-        checker_params["max_error"] = str(max_traces)
+    if max_trace:
+        checker_params["max_error"] = str(max_trace)
 
     model = Model.parse_file(str(model_path), init, next)
     return model.sample(

--- a/atomkraft/test/__init__.py
+++ b/atomkraft/test/__init__.py
@@ -90,6 +90,11 @@ def model(
     max_trace: Optional[int] = typer.Option(
         None, show_default=False, help="Maximum number of traces to generate"
     ),
+    view: Optional[str] = typer.Option(
+        None,
+        show_default=False,
+        help="View projector to generate only interesting traces",
+    ),
     verbose: bool = typer.Option(
         False, "--verbose", "-v", help="Output logging on console"
     ),
@@ -99,7 +104,14 @@ def model(
     """
     tests = [t.strip() for ts in test for t in ts.split(",")]
 
-    exit_code = test_model(model, tests, reactor, keypath, max_trace, verbose)
+    checker_args = dict()
+
+    if max_trace:
+        checker_args["max_error"] = str(max_trace)
+    if view:
+        checker_args["view"] = view
+
+    exit_code = test_model(model, tests, reactor, keypath, checker_args, verbose)
 
     if model:
         with AtomkraftConfig() as c:

--- a/atomkraft/test/__init__.py
+++ b/atomkraft/test/__init__.py
@@ -87,7 +87,7 @@ def model(
         show_default=True,
         help="Path to key used as step name, extracted from ITF states",
     ),
-    max_traces: Optional[int] = typer.Option(
+    max_trace: Optional[int] = typer.Option(
         None, show_default=False, help="Maximum number of traces to generate"
     ),
     verbose: bool = typer.Option(
@@ -99,7 +99,7 @@ def model(
     """
     tests = [t.strip() for ts in test for t in ts.split(",")]
 
-    exit_code = test_model(model, tests, reactor, keypath, max_traces, verbose)
+    exit_code = test_model(model, tests, reactor, keypath, max_trace, verbose)
 
     if model:
         with AtomkraftConfig() as c:

--- a/atomkraft/test/model.py
+++ b/atomkraft/test/model.py
@@ -3,7 +3,7 @@ import shutil
 import time
 from datetime import datetime
 from pathlib import Path
-from typing import List, Optional
+from typing import Dict, List, Optional
 
 import pytest
 from atomkraft.config.atomkraft_config import AtomkraftConfig
@@ -48,7 +48,7 @@ def test_model(
     tests: List[str],
     reactor: Optional[Path],
     keypath: str,
-    max_trace: Optional[int],
+    checker_params: Dict[str, str],
     verbose: bool,
 ) -> int:
     """
@@ -72,7 +72,9 @@ def test_model(
     print(f"Generating traces for {model.name} ...")
 
     try:
-        model_result = generate_traces(None, model, tests, max_trace=max_trace)
+        model_result = generate_traces(
+            None, model, tests, checker_params=checker_params
+        )
     except Exception as e:
         raise RuntimeError(f"[Modelator] {e}")
 

--- a/atomkraft/test/model.py
+++ b/atomkraft/test/model.py
@@ -91,28 +91,26 @@ def test_model(
         .replace("-", "_")
     )
 
-    root_report_dir = root / "reports" / test_group
+    test_name = f"test_{test_group}"
 
     successul_ops = model_result.successful()
 
-    exit_codes = []
+    test_list = []
 
     for op in successul_ops:
-        print(f"Testing {op} ...")
+        print(f"Preparing test for {op} ...")
         for trace_path in model_result.trace_paths(op):
             trace = Path(trace_path)
             if all(not c.isdigit() for c in trace.name):
                 continue
 
             trace_name = trace.name.removesuffix(".itf.json")
-
-            print(trace)
+            print(f"Using {trace} ...")
             trace = get_relative_project_path(trace)
-            test_name = f"test_{test_group}"
             test_path = test_dir / test_name / f"test_{op}_{trace_name}.py"
             test_path.parent.mkdir(parents=True, exist_ok=True)
             with open(test_path, "w") as test:
-                print(f"Writing {test_name} ...")
+                print(f"Writing {test_path.name} ...")
                 test.write(
                     TRACE_TEST_STUB.format(
                         json.dumps(str(reactor).replace("/", ".").removesuffix(".py")),
@@ -120,31 +118,35 @@ def test_model(
                         json.dumps(keypath),
                     )
                 )
+            test_list.append((trace, test_path))
 
-            report_dir = root_report_dir / test_path.stem
-            report_dir.mkdir(parents=True, exist_ok=True)
+    report_dir = root / "reports" / test_name
+    report_dir.mkdir(parents=True, exist_ok=True)
 
-            logging_file = report_dir / "log.txt"
+    logging_file = report_dir / "log.txt"
 
-            pytest_report_file = report_dir / "report.jsonl"
+    pytest_report_file = report_dir / "report.jsonl"
 
-            pytest_args = [
-                "--log-file-level=INFO",
-                "--log-cli-level=INFO",
-                f"--log-file={logging_file}",
-                f"--report-log={pytest_report_file}",
-            ]
+    pytest_args = [
+        "--log-file-level=INFO",
+        "--log-cli-level=INFO",
+        f"--log-file={logging_file}",
+        f"--report-log={pytest_report_file}",
+    ]
 
-            if verbose:
-                pytest_args.append("-rP")
+    if verbose:
+        pytest_args.append("-rP")
 
-            exit_codes.append(pytest.main(pytest_args + [str(test_path)]))
+    exit_code = pytest.main(
+        pytest_args + [str(test_file) for (_, test_file) in test_list]
+    )
 
-            copy_if_exists([Path(trace), root / ".atomkraft" / "nodes"], report_dir)
+    for (trace, _) in test_list:
+        copy_if_exists([Path(trace), root / ".atomkraft" / "nodes"], report_dir)
 
     if successul_ops:
-        print(f"Test data is saved at {root_report_dir}")
-        return max(int(e) for e in exit_codes)
+        print(f"Test data is saved at {report_dir}")
     else:
         print("No trace is produced.")
-        return 0
+
+    return int(exit_code)

--- a/atomkraft/test/model.py
+++ b/atomkraft/test/model.py
@@ -48,7 +48,7 @@ def test_model(
     tests: List[str],
     reactor: Optional[Path],
     keypath: str,
-    max_traces: Optional[int],
+    max_trace: Optional[int],
     verbose: bool,
 ) -> int:
     """
@@ -72,7 +72,7 @@ def test_model(
     print(f"Generating traces for {model.name} ...")
 
     try:
-        model_result = generate_traces(None, model, tests, max_traces=max_traces)
+        model_result = generate_traces(None, model, tests, max_trace=max_trace)
     except Exception as e:
         raise RuntimeError(f"[Modelator] {e}")
 

